docs/architecture/adr/ADR040-cluster-stability-node-replacement.md

```diff
@@ -0,0 +1,137 @@
+
+Pros:
+* AWS should mostly take care of the process for us, we'd just need some code to call SignalResource.
+* The cluster should not encounter any health issues if new instances are ready to accept new pods before existing instances get terminated
+
+Cons:
+* Our Concourse job would probably time out while this process happens, but we can probably adjust/live with that.
+
+### Action 2
+
+Remove this CloudFormation UpdatePolicy and find some alternative to the update policy that ensures the cluster eventually fully rolls out the expected launch template. Mutually exclusive with action 2.
+
+#### Action 2A
+
+Manual cordoning, draining and termination of nodes
+
+Pros:
+* 
+
+Cons:
+* This would make it some rather irritating toil as an engineer would have to manually run the process.
+* It might be possible to make a mistake here and cordon/drain too many nodes at once without termination so therefore no replacement.
+
+#### Action 2B
+
+A script that we run that cordons, drains and terminates nodes - triggered manually.
+
+Pros:
+* Would be able to choose exactly when we replace nodes to optimise for low load, no other critical operations ongoing, staff availability in the event of a problem, etc.
+
+Cons:
+* We'd have to notice the new launch template and trigger it manually.
+
+#### Action 2C
+
+A script that runs regularly that terminates nodes - this would ensure the cluster is eventually updated, just not immediately.
+
+Pros:
+* Has the added benefit of regularly killing off nodes regardless of launch template changes, meaning nodes don't stick around for too long.
+
+Cons:
+* If this was a CronJob inside the cluster, it might terminate itself while running and leave us with an inconsistent cluster? But if it's a Lambda maybe not. Or if it's written to only kill one node and wait until the next run to kill another we might also be okay.
+* Would not immediately replace all the nodes, rollout would be slow.
+* If the new launch template is broken, the cluster would slowly break itself through terminating the remaining healthy nodes.
+
+### Action 2D
+
+Write an operator that manages the ASGs for the worker nodes that works with / replaces the cluster autoscaler. Mutually exclusive with action 3.
+
+Pros:
+* Would be able to manage both node replacement for a new launch template and scale for load at the same time.
+
+Cons:
+* More code for us to write and maintain.
+* Potentially re-inventing the wheel if we have to replace the cluster autoscaler.
+
+### Action 3
+
+A lifecycle hook in the ASG that tells it to cordon and drain nodes etc. when it terminates them.
+
+Pros:
+* This might have some benefit if we want to e.g. use spot instances in the future, as we'd be confident we can quickly cordon and drain a node in the event of a notification saying a current worker node will be terminated shortly.
+
+Cons:
+* Nodes might still get terminated too quickly for the cluster to cope.
```

Comment:
I don't think this is true? Lifecycle hooks allow you to pause the ASG activity until the node is safely drained. You send an event saying that you're ready for the ASG to continue and then it carries on.  If you take longer than an hour to drain though you might see issues:

> By default, the instance remains in a wait state for one hour, and then the Auto Scaling group continues the launch or terminate process (Pending:Proceed or Terminating:Proceed). If you need more time, you can restart the timeout period by recording a heartbeat. If you finish before the timeout period ends, you can complete the lifecycle action, which continues the launch or termination process.